### PR TITLE
Fixes #1672 Errors are not propagated on failed simulation run

### LINKED ISSUE #1672: No error is propagated when `SimulationBatch` run returns an error
See description in https://github.com/Open-Systems-Pharmacology/OSPSuite.SimModel/issues/146 - it seems that the error is thrown silently and is not propagated.

@abdelr mayber you can tell more where the error is actually thrown?

### PR DESCRIPTION
**Pull request by rwmcintosh**

### CODE REVIEW

File: src/OSPSuite.R/Domain/SimulationBatch.cs [hunk 1]
```diff
@@ -148,12 +149,15 @@ public SimulationResults Run(SimulationBatchRunValues simulationBatchRunValues)
       {
          _simModelBatch.UpdateParameterValues(simulationBatchRunValues.Values);
          _simModelBatch.UpdateInitialValues(simulationBatchRunValues.MoleculeValues);
-         var simulationResults = _simModelBatch.RunSimulation();
+         var simulationRunResults = _simModelBatch.RunSimulation();
+
+         if (!simulationRunResults.Success)
+            throw new OSPSuiteException(simulationRunResults.Error);
```
Comment (rwmcintosh) [inline src/OSPSuite.R/Domain/SimulationBatch.cs:155]: Unsuccessful run must be signaled by exception

File: src/OSPSuite.Core/Domain/Services/SimModelBatch.cs [hunk 2]
```diff
@@ -79,7 +80,10 @@ public SimulationRunResults RunSimulation()
          {
             updateSimulationValues();
             _simModelSimulation.RunSimulation();
-            return new SimulationRunResults(success: true, warnings: WarningsFrom(_simModelSimulation), results: getResults());
+            var hasResults = simulationHasResults(_simModelSimulation);
+
+            return new SimulationRunResults(success: hasResults, warnings: WarningsFrom(_simModelSimulation), results: getResults(),
+               error: hasResults ? null : Error.SimulationDidNotProduceResults);
```
Comment (msevestre) [inline src/OSPSuite.Core/Domain/Services/SimModelBatch.cs:86]: why don't you create a method (something around the lines of...)
```
createSimulationResults(string errorFromException = null) {
var hasResults = simulationHasResults(_simModelSimulation);
var warnings = WarningsFrom(_simModelSimulation);
var error = errorFromException ??  (hasResults ? null : Error.SimulationDidNotProduceResults)
if(error==null){
return SimulationRunResults(success: true , warnings, results: getResults())
}

//error
return SimulationRunResults(false, warnings, error)
}
```
and call that in both instances?

File: src/OSPSuite.R/Domain/SimulationBatch.cs (hunk 1 above)
Comment (msevestre) [reply src/OSPSuite.R/Domain/SimulationBatch.cs:155]: yep

File: tests/OSPSuite.R.Tests/Services/ConcurrentSimulationRunnerSpecs.cs [hunk 3]
```diff
@@ -87,6 +88,66 @@ public void should_complete_without_error()
...
+         _results = sut.RunConcurrently();
+      }
+
+      [Observation]
+      public void should_throw_exception_when_the_simulation_run_fails()
+      {
+         var id = _ids.First();
+         var simulationBatch = Api.GetSimulationBatchFactory().Create(_simulation, _concurrentRunSimulationBatch.SimulationBatchOptions);
+            
+         The.Action(() => simulationBatch.Run(_simulationBatchRunValues.FirstOrDefault(v => v.Id == id)))
```
Comment (msevestre) [inline tests/OSPSuite.R.Tests/Services/ConcurrentSimulationRunnerSpecs.cs:146]: hum I don't get it. The action was called already in RunConcurrently
So i think this test should check that _results has en entry with success=false and an error

That tests here is a test for the _simulationBatch or sthg. No?

**Review by msevestre — CHANGES_REQUESTED** [review]
I am wondering if we can have the implementation only once for the returning of the results
Also the tests is not what I would except. I feel you are testing something else (this test is valid  but for another sut)

File: src/OSPSuite.Core/Domain/Services/SimModelBatch.cs [hunk 4]
```diff
@@ -93,6 +95,25 @@ public SimulationRunResults RunSimulation()
...
+         {
+            return new SimulationRunResults(success: true, warnings, results: getResults());
+         }
+
+         //error
+         return new SimulationRunResults(false, warnings, new DataRepository(), error);
+      }
+
+      private bool simulationHasResults(Simulation simModelSimulation)
+      {
```
Comment (msevestre) [inline src/OSPSuite.Core/Domain/Services/SimModelBatch.cs:113]: FYI: This is ok to use the short hand syntax (see below)


```
private bool simulationHasResults(Simulation simModelSimulation) => simModelSimulation.AllValues.Any();
```

File: tests/OSPSuite.R.Tests/Services/ConcurrentSimulationRunnerSpecs.cs [hunk 5]
```diff
@@ -87,6 +88,64 @@ public void should_complete_without_error()
...
+         _concurrentRunSimulationBatch.AddSimulationBatchRunValues(_simulationBatchRunValues[0]);
+
+         _ids.AddRange(_concurrentRunSimulationBatch.SimulationBatchRunValues.Select(x => x.Id));
+         _results = sut.RunConcurrently();
+      }
+
+      [Observation]
+      public void should_return_a_result_with_success_set_to_false()
+      {
+         _results.All(x => x.Succeeded).ShouldBeFalse();
```
Comment (msevestre) [inline tests/OSPSuite.R.Tests/Services/ConcurrentSimulationRunnerSpecs.cs:145]: Do we have a test that now verifies that this flag is sometimes true?  (e.g. can you set the parameter for one run instead as for all). Then one run should be success and the one where it fails should be false.

Comment (rwmcintosh) [reply tests/OSPSuite.R.Tests/Services/ConcurrentSimulationRunnerSpecs.cs]: We do now.

**Pull request merged**